File: ios/GradeThread/Passport/PassportTypes.swift
import Foundation

/// PII-free passport timeline as returned by the public edge endpoint
/// `GET /api/passport/:slug` (services/edge-functions/src/routes/passport.ts).
/// The edge already sanitizes payloads and resolves actors to pseudonymous
/// labels only — iOS renders what it is given and never reaches for identity.
///
/// Decoded with ``JSONDecoder/iso8601`` (the EdgeAPI decoder), so snake_case
/// keys arrive camelCased here. Timestamps are kept as raw ISO strings (the
/// edge emits fractional seconds, which the default `.iso8601` date strategy
/// rejects) and formatted at the display boundary — mirroring the web page.
struct PassportTimeline: Decodable, Equatable {
    let slug: String
    let skuClass: [String: PassportJSONValue]
    let status: String
    let createdAt: String
    /// PII-free origin-seller Verified badge (null unless the seller opted in).
    let originVerifiedSeller: PassportVerifiedSeller?
    let events: [PassportEvent]

    private enum CodingKeys: String, CodingKey {
        case slug, skuClass, status, createdAt, originVerifiedSeller, events
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        slug = try c.decode(String.self, forKey: .slug)
        skuClass = (try? c.decode([String: PassportJSONValue].self, forKey: .skuClass)) ?? [:]
        status = (try? c.decode(String.self, forKey: .status)) ?? "active"
        createdAt = (try? c.decode(String.self, forKey: .createdAt)) ?? ""
        originVerifiedSeller = try? c.decodeIfPresent(PassportVerifiedSeller.self, forKey: .originVerifiedSeller)
        events = (try? c.decode([PassportEvent].self, forKey: .events)) ?? []
    }

    /// Test/preview constructor.
    init(
        slug: String,
        skuClass: [String: PassportJSONValue] = [:],
        status: String = "active",
        createdAt: String = "",
        originVerifiedSeller: PassportVerifiedSeller? = nil,
        events: [PassportEvent] = []
    ) {
        self.slug = slug
        self.skuClass = skuClass
        self.status = status
        self.createdAt = createdAt
        self.originVerifiedSeller = originVerifiedSeller
        self.events = events
    }
}

/// One hop on the provenance chain.
struct PassportEvent: Decodable, Equatable, Identifiable {
    let eventType: String
    let confidence: String
    /// Pseudonymous actor label (e.g. "Seller A"), or nil for an actor-less hop.
    let actor: String?
    /// The actor's PUBLIC Verified identity, present ONLY when they opted in for
    /// this hop. nil (the default) keeps the actor pseudonymous.
    let actorRevealed: PassportVerifiedSeller?
    let source: String?
    let payload: PassportEventPayload
    let createdAt: String

    /// Stable identity for `ForEach` (the ledger is append-only, so type+time+
    /// actor is unique enough; index disambiguates same-instant duplicates).
    var id: String { "\(eventType)|\(createdAt)|\(actor ?? "")" }

    private enum CodingKeys: String, CodingKey {
        case eventType, confidence, actor, actorRevealed, source, payload, createdAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        eventType = (try? c.decode(String.self, forKey: .eventType)) ?? "unknown"
        confidence = (try? c.decode(String.self, forKey: .confidence)) ?? "unknown"
        actor = try? c.decodeIfPresent(String.self, forKey: .actor)
        actorRevealed = try? c.decodeIfPresent(PassportVerifiedSeller.self, forKey: .actorRevealed)
        source = try? c.decodeIfPresent(String.self, forKey: .source)
        payload = (try? c.decode(PassportEventPayload.self, forKey: .payload)) ?? PassportEventPayload()
        createdAt = (try? c.decode(String.self, forKey: .createdAt)) ?? ""
    }

    init(
        eventType: String,
        confidence: String,
        actor: String? = nil,
        actorRevealed: PassportVerifiedSeller? = nil,
        source: String? = nil,
        payload: PassportEventPayload = PassportEventPayload(),
        createdAt: String = ""
    ) {
        self.eventType = eventType
        self.confidence = confidence
        self.actor = actor
        self.actorRevealed = actorRevealed
        self.source = source
        self.payload = payload
        self.createdAt = createdAt
    }
}

/// The subset of a (PII-free) event payload iOS surfaces. All optional — a
/// `graded` event carries the grade detail + a public certificate handle; other
/// event types carry none of these (the keys simply decode to nil).
struct PassportEventPayload: Decodable, Equatable {
    let overallScore: Double?
    let gradeTier: String?
    let certificate: String?

    init(overallScore: Double? = nil, gradeTier: String? = nil, certificate: String? = nil) {
        self.overallScore = overallScore
        self.gradeTier = gradeTier
        self.certificate = certificate
    }
}

/// A public, opt-in Verified profile (handle + display name only — never a user
/// id or email). Used for both the origin seller badge and a per-hop reveal.
struct PassportVerifiedSeller: Decodable, Equatable {
    let handle: String
    let displayName: String?
}

/// A minimal JSON value so `sku_class` (a free-form descriptor) decodes without
/// constraining its shape. Only the string-bearing fields are read for display.
enum PassportJSONValue: Decodable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: PassportJSONValue])
    case array([PassportJSONValue])
    case null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
        } else if let v = try? c.decode(Bool.self) {
            self = .bool(v)
        } else if let v = try? c.decode(Double.self) {
            self = .number(v)
        } else if let v = try? c.decode(String.self) {
            self = .string(v)
        } else if let v = try? c.decode([String: PassportJSONValue].self) {
            self = .object(v)
        } else if let v = try? c.decode([PassportJSONValue].self) {
            self = .array(v)
        } else {
            self = .null
        }
    }

    /// The string value if this is a non-empty string, else nil.
    var stringValue: String? {
        if case let .string(s) = self, !s.trimmingCharacters(in: .whitespaces).isEmpty {
            return s
        }
        return nil
    }
}

// MARK: - Presentation helpers (pure)

extension PassportEvent {
    /// Human label + SF Symbol per event type. Mirrors the web `EVENT_META`
    /// map; the set tracks the `garment_event_type` enum (00256).
    var presentation: (label: String, symbol: String) {
        switch eventType {
        case "graded": return ("Condition graded", "checkmark.seal.fill")
        case "listed": return ("Listed for sale", "tag.fill")
        case "sold": return ("Sold", "bag.fill")
        case "ownership_transfer": return ("Ownership transferred", "arrow.left.arrow.right")
        case "fingerprinted": return ("Fingerprinted", "viewfinder")
        default: return (PassportFormat.titleCase(eventType), "clock.arrow.circlepath")
        }
    }
}

/// Small pure formatters shared by the passport view.
enum PassportFormat {
    /// "ownership_transfer" / "very-good" → "Ownership Transfer" / "Very Good".
    static func titleCase(_ value: String) -> String {
        value
            .split(whereSeparator: { $0 == "-" || $0 == "_" })
            .map { $0.prefix(1).uppercased() + $0.dropFirst() }
            .joined(separator: " ")
    }

    /// A human garment name from the PII-free `sku_class` descriptor. The
    /// EdgeAPI decoder camelCases nested keys (`garment_type` → `garmentType`),
    /// so look up both forms to stay robust to the transport.
    static func garmentName(from sku: [String: PassportJSONValue]) -> String {
        let brand = sku["brand"]?.stringValue
        let rawType = (sku["garmentType"] ?? sku["garment_type"])?.stringValue
        let type = rawType.map { titleCase($0) }
        let parts = [brand, type].compactMap { $0 }
        return parts.isEmpty ? "Graded garment" : parts.joined(separator: " ")
    }

    /// ISO timestamp → "June 20, 2026" (falls back to the raw string).
    static func longDate(_ iso: String) -> String {
        guard !iso.isEmpty else { return "" }
        let parsers: [ISO8601DateFormatter] = {
            let withFraction = ISO8601DateFormatter()
            withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let plain = ISO8601DateFormatter()
            plain.formatOptions = [.withInternetDateTime]
            return [withFraction, plain]
        }()
        for parser in parsers {
            if let date = parser.date(from: iso) {
                let out = DateFormatter()
                out.dateStyle = .long
                out.timeStyle = .none
                return out.string(from: date)
            }
        }
        return iso
    }
}

File: android/app/src/main/java/com/gradethread/app/passport/PassportTypes.kt
package com.gradethread.app.passport

import androidx.annotation.StringRes
import com.gradethread.app.R
import com.gradethread.app.ui.UiMessage

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale

/**
 * US-1376 (iOS `PassportTypes`): the PII-free provenance chain from
 * `GET /api/passport/{slug}`.
 *
 * The edge sanitizes everything before it leaves: actors are pseudonymous
 * labels unless the owner opted in per hop, and no user id or email is ever in
 * the payload. This client renders what it is given and never reaches for
 * identity — the privacy promise is the edge's, and the way to keep it is to
 * have nothing here that could break it.
 */
@Serializable
data class PassportTimeline(
    val slug: String = "",
    /** Free-form garment descriptor; only its string fields are read. */
    @SerialName("sku_class") val skuClass: JsonObject = JsonObject(emptyMap()),
    val status: String = "active",
    @SerialName("created_at") val createdAt: String = "",
    @SerialName("origin_verified_seller") val originVerifiedSeller: PassportVerifiedSeller? = null,
    val events: List<PassportEvent> = emptyList(),
)

/** One hop on the chain. */
@Serializable
data class PassportEvent(
    @SerialName("event_type") val eventType: String = "unknown",
    val confidence: String = "unknown",
    /** Pseudonymous label such as "Seller A", or null for an actor-less hop. */
    val actor: String? = null,
    /** The actor's PUBLIC identity, present only when they opted in for this hop. */
    @SerialName("actor_revealed") val actorRevealed: PassportVerifiedSeller? = null,
    val source: String? = null,
    val payload: PassportEventPayload = PassportEventPayload(),
    @SerialName("created_at") val createdAt: String = "",
) {
    /**
     * Stable list key.
     *
     * The ledger is append-only, so type + timestamp + actor is unique enough;
     * the index the caller adds disambiguates two hops recorded in the same
     * instant.
     */
    val key: String get() = "$eventType|$createdAt|${actor.orEmpty()}"
}

/** The subset of a sanitized payload this surface shows. */
@Serializable
data class PassportEventPayload(
    @SerialName("overall_score") val overallScore: Double? = null,
    @SerialName("grade_tier") val gradeTier: String? = null,
    val certificate: String? = null,
)

/** A public, opt-in Verified profile. Handle and display name only. */
@Serializable
data class PassportVerifiedSeller(
    val handle: String = "",
    @SerialName("display_name") val displayName: String? = null,
    val since: String? = null,
)

/**
 * The confidence taxonomy, mirroring `src/lib/passport-confidence.ts` and the
 * `garment_event_confidence` enum. Exactly three levels.
 *
 * The wording is deliberately measured. "Probable" and "unknown" never say
 * confirmed or guaranteed, because a passport is a provenance claim someone
 * may rely on when paying, and overstating a link is the one failure that
 * matters here.
 */
enum class PassportConfidence(@StringRes val label: Int, @StringRes val explanation: Int) {
    DETERMINISTIC(
        R.string.passport_confidence_verified,
        R.string.passport_confidence_verified_why,
    ),
    PROBABLE(
        R.string.passport_confidence_probable,
        R.string.passport_confidence_probable_why,
    ),
    UNKNOWN(
        R.string.passport_confidence_unverified,
        R.string.passport_confidence_unverified_why,
    ),
    ;

    companion object {
        /** Anything unrecognized reads as UNKNOWN, never as verified. */
        fun of(raw: String?): PassportConfidence = when (raw) {
            "deterministic" -> DETERMINISTIC
            "probable" -> PROBABLE
            else -> UNKNOWN
        }
    }
}

/** How many hops are proven versus inferred. Mirrors the web `chainStrength()`. */
data class PassportChainStrength(
    val total: Int,
    val deterministic: Int,
    val probable: Int,
    val unknown: Int,
    /** Fraction of links that are deterministic, 0..1. */
    val score: Double,
    @StringRes val label: Int,
    val summary: UiMessage,
) {
    companion object {
        fun of(confidences: List<String?>): PassportChainStrength {
            var det = 0
            var prob = 0
            var unk = 0
            for (raw in confidences) {
                when (PassportConfidence.of(raw)) {
                    PassportConfidence.DETERMINISTIC -> det++
                    PassportConfidence.PROBABLE -> prob++
                    PassportConfidence.UNKNOWN -> unk++
                }
            }
            val total = confidences.size
            val score = if (total == 0) 0.0 else det.toDouble() / total
            return PassportChainStrength(
                total = total,
                deterministic = det,
                probable = prob,
                unknown = unk,
                score = score,
                label = when {
                    total == 0 -> R.string.passport_strength_none
                    score >= 0.75 -> R.string.passport_strength_strong
                    score >= 0.4 -> R.string.passport_strength_moderate
                    else -> R.string.passport_strength_emerging
                },
                summary = if (total == 0) {
                    UiMessage(R.string.passport_no_history)
                } else {
                    // Counts, not an adjective on its own: "3 of 5 verified" is
                    // checkable, "moderate" is a word somebody has to trust.
                    //
                    // US-2976: pluralised on the TOTAL, because "links are" is
                    // agreeing with the five, not with the three.
                    UiMessage.plural(
                        R.plurals.passport_verified_links,
                        args = listOf(det, total),
                        quantity = total,
                    )
                },
            )
        }
    }
}

object PassportFormat {

    /**
     * Human label per event type, mirroring the web `EVENT_META`.
     *
     * `authenticity_assessed` is here on purpose: the enum gained it in
     * migration 00488 and the iOS map still doesn't have it, so that event
     * falls through to a generic title-cased label there.
     */
    fun eventLabel(eventType: String): UiMessage = when (eventType) {
        "graded" -> UiMessage(R.string.passport_event_graded)
        "listed" -> UiMessage(R.string.passport_event_listed)
        "sold" -> UiMessage(R.string.passport_event_sold)
        "ownership_transfer" -> UiMessage(R.string.passport_event_ownership_transfer)
        "fingerprinted" -> UiMessage(R.string.passport_event_fingerprinted)
        "authenticity_assessed" -> UiMessage(R.string.passport_event_authenticity_assessed)
        // A future enum value renders as itself rather than as "Unknown", which
        // would hide a real event behind a word that means nothing happened.
        //
        // US-2976: it rides as `detail`, so it is shown exactly as the server
        // named it. Untranslated is the honest outcome for a word we have never
        // seen; inventing a resource for it would not be.
        else -> UiMessage(R.string.passport_event_other, detail = titleCase(eventType))
    }

    /** "ownership_transfer" / "very-good" → "Ownership Transfer" / "Very Good". */
    fun titleCase(value: String): String = value
        .split('-', '_')
        .filter { it.isNotEmpty() }
        .joinToString(" ") { it.replaceFirstChar { c -> c.uppercaseChar() } }

    /** A human garment name from the PII-free descriptor. */
    fun garmentName(skuClass: JsonObject): String {
        val brand = string(skuClass, "brand")
        val type = string(skuClass, "garment_type")?.let { titleCase(it) }
        val parts = listOfNotNull(brand, type)
        return if (parts.isEmpty()) "Graded garment" else parts.joinToString(" ")
    }

    private fun string(obj: JsonObject, key: String): String? = (obj[key] as? JsonPrimitive)
        ?.takeIf { it.isString }
        ?.content
        ?.trim()
        ?.takeIf { it.isNotEmpty() }

    /**
     * ISO timestamp to a readable date, falling back to the raw string.
     *
     * Never blank on a parse failure: showing the raw timestamp is ugly but
     * still tells the reader when something happened, which an empty line does
     * not.
     */
    fun longDate(iso: String): String {
        if (iso.isBlank()) return ""
        return runCatching {
            OffsetDateTime.parse(iso)
                .format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG).withLocale(Locale.getDefault()))
        }.getOrElse { iso }
    }

    /** Sort key for chronological ordering; unparseable timestamps sink. */
    fun epochMillis(iso: String): Long = runCatching { OffsetDateTime.parse(iso).toInstant().toEpochMilli() }
        .getOrDefault(Long.MAX_VALUE)

    /**
     * Chronological order, oldest first.
     *
     * The edge already orders these, but it also CACHES the assembled body, and
     * a timeline that renders out of order tells a false story about what
     * happened to a garment. Sorting again costs nothing.
     */
    fun ordered(events: List<PassportEvent>): List<PassportEvent> = events.sortedBy { epochMillis(it.createdAt) }

    /** The grade line on a `graded` hop, or null when it carries none. */
    fun gradeLine(event: PassportEvent): String? {
        val score = event.payload.overallScore ?: return null
        val tier = event.payload.gradeTier?.takeIf { it.isNotBlank() }?.let { " · ${titleCase(it)}" }
        return String.format(Locale.US, "%.1f", score) + tier.orEmpty()
    }

    /**
     * Who did it.
     *
     * A revealed handle wins, then the pseudonymous label, then nothing —
     * inventing "Unknown seller" for an actor-less hop would imply a person was
     * involved when the ledger says otherwise.
     */
    fun actorLine(event: PassportEvent): String? {
        event.actorRevealed?.let { revealed ->
            val name = revealed.displayName?.takeIf { it.isNotBlank() } ?: "@${revealed.handle}"
            return "$name (verified)"
        }
        return event.actor?.takeIf { it.isNotBlank() }
    }
}
